limitador/src/limit/cel.rs

```diff
@@ -0,0 +1,370 @@
+use crate::limit::Limit;
+use cel_interpreter::{ExecutionError, Value};
+use serde::{Deserialize, Serialize};
+use std::cmp::Ordering;
+use std::collections::{HashMap, HashSet};
+use std::hash::{Hash, Hasher};
+use std::sync::Arc;
+
+pub use errors::{EvaluationError, ParseError};
+
+pub(super) mod errors {
+    use cel_interpreter::ExecutionError;
+    use std::error::Error;
+    use std::fmt::{Display, Formatter};
+
+    #[derive(Debug, PartialEq)]
+    pub enum EvaluationError {
+        UnexpectedValueType(String),
+        ExecutionError(ExecutionError),
+    }
+
+    impl Display for EvaluationError {
+        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+            match self {
+                EvaluationError::UnexpectedValueType(value) => {
+                    write!(f, "unexpected value of type {}", value)
+                }
+                EvaluationError::ExecutionError(error) => error.fmt(f),
+            }
+        }
+    }
+
+    impl Error for EvaluationError {
+        fn source(&self) -> Option<&(dyn Error + 'static)> {
+            match self {
+                EvaluationError::UnexpectedValueType(_) => None,
+                EvaluationError::ExecutionError(err) => Some(err),
+            }
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct ParseError {
+        input: String,
+        source: Box<dyn Error + 'static>,
+    }
+
+    impl ParseError {
+        pub fn from(source: cel_parser::ParseError, input: String) -> Self {
+            Self {
+                input,
+                source: Box::new(source),
+            }
+        }
+    }
+
+    impl Display for ParseError {
+        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+            write!(f, "couldn't parse {}: {}", self.input, self.source)
+        }
+    }
+
+    impl Error for ParseError {
+        fn source(&self) -> Option<&(dyn Error + 'static)> {
+            Some(self.source.as_ref())
+        }
+    }
+
+    impl From<ExecutionError> for EvaluationError {
+        fn from(err: ExecutionError) -> Self {
+            EvaluationError::ExecutionError(err)
+        }
+    }
+}
+
+pub struct Context<'a> {
+    variables: HashSet<String>,
+    ctx: cel_interpreter::Context<'a>,
+}
+
+impl Context<'_> {
+    pub(crate) fn new(limit: &Limit, root: String, values: &HashMap<String, String>) -> Self {
```

Comment:
So, this `root` relates to this point from #385

>Should the descriptor(s) and their entries be first class citizens and have their own root binding?

Right now, `limit` binds the limit info (`id` & `name`), I was thinking that possibly we could have some root binding for the descriptor entries (e.g. `descriptor`), which... could then also be a `List` and hold multiple of them and all their entries: `descriptor[0].foo == '1' && descriptor[1].bar != 'bad'`... See [this issue here](https://github.com/Kuadrant/limitador/issues/62), so it'd probably be a `List<HashMap<String, String>>` as well then...

Whatever the case, who would set that `root` namespace/binding to something? I'd expect it to be the server code... but that then creates further issues in the (de)serialization paths... Also, the logic of resolving the `Limit.variables()` would need to account for that (along side what happens in the wasm-shim's code to lazily resolve attributes from the host).

Reply:
Also, doing so would tie releasing this with a release of the kuadrant-operator... right now, this isn't necessary.

Reply:
So... thinking about this a little more, and given [this comment](https://github.com/Kuadrant/limitador/pull/396#discussion_r1865738487), here's how I'm proposing to go about this:

- [ ] Have a blanket `From<HashMap<String, String>` for `Context`... which would _not_ add the binding for `limit`.
- [ ] Make `Context::new` public to all users
- [ ] Change `Limit::applies` & `::resolve_variables` to take a `&Context` instead of creating it explicitly
- [ ] Have the `(Async)RateLimiter::counters_that_apply` also take the `&Context`
- [ ] Have the `(Async)RateLimiter` APIs expose the `Context`, instead of `&HashMap<String, String>` 

That way, the server code can decide how it wires the `RateLimitDescriptor`s and their `.entries` from envoy's protocol for external rate limiting.

Reply:
> Have the (Async)RateLimiter::counters_that_apply also take the &Context
> Have the (Async)RateLimiter APIs expose the Context, instead of &HashMap<String, String>

duh! this obviously doesn't work 🤦 Or at least not while injecting the `Limit` this operates on.